src/vsr/journal.zig

```diff
@@ -703,83 +843,505 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
+            const self = read.self;
+            const replica = @fieldParentPtr(Replica, "journal", self);
+            assert(!self.recovered);
+            assert(self.recovering);
+            assert(read.destination_replica == null);
+
+            const slot = Slot{ .index = @intCast(u64, read.checksum) };
+            assert(slot.index < slot_count);
+
+            // `recover_headers_on_read` either sets both dirty+faulty, or neither.
+            if (self.dirty.bit(slot)) {
+                assert(self.faulty.bit(slot));
+            }
+
+            const Decision = enum {
+                /// The header and prepare are identical; no repair necessary.
+                eql,
+                /// Reserved; clear dirty/faulty, no repair necessary.
+                nil,
+                /// If replica_count>1: Repair with VSR `request_prepare`. Mark dirty, clear faulty.
+                /// If replica_count=1: Use intact prepare. Clear dirty, clear faulty.
+                fix,
+                /// If replica_count>1: Repair with VSR `request_prepare`. Mark dirty and faulty.
+                /// If replica_count=1: Fail; cannot recover safely.
+                vsr,
+            };
+
+            const header = &self.headers[slot.index];
+            const header_reserved = header.command == .reserved;
+            const header_valid = !self.faulty.bit(slot);
+            assert(header.invalid() == null);
+
+            const prepare = read.message.header;
+            const prepare_checksum = prepare.valid_checksum();
+            // A prepare in the wrong cluster/slot indicates a misdirected read or write.
+            const prepare_cluster = prepare.cluster == replica.cluster;
+            const prepare_slot = switch (prepare.command) {
+                .prepare => slot.index == self.slot_for_op(prepare.op).index,
+                .reserved => slot.index == prepare.op,
+                else => false,
+            };
+            // All journalled messages should reserved or prepares.
+            // A misdirected read/write to/from the block zone may return the wrong type of message.
+            const prepare_reserved = prepare.command == .reserved;
+            const prepare_prepare = prepare.command == .prepare;
+            const prepare_valid = prepare_checksum and prepare_cluster and prepare_slot and
+                (prepare_reserved or prepare_prepare);
+            if (prepare_valid) assert(prepare.invalid() == null);
+
+            const match_checksum = header.checksum == prepare.checksum;
+            const match_op = header.op == prepare.op;
+            const match_view = header.view == prepare.view;
+
+            const action: struct {
+                label: []const u8,
+                decision: Decision,
+            } = action: {
+                if (!header_valid and !prepare_valid) {
+                    break :action .{ .label = "@A", .decision = .vsr };
+                }
```

Comment:
Let's use a separate function to return the action struct.

This will decouple the decision from the I/O callback for nice unit testing, help to keep the function body under 70 lines, and help to thin out the control flow lines to `return .{ .label = "@A", .decision = .vsr };`

Can we also further break this function into 2/3 sub-functions (as makes sense) so we can keep the function bodies under 70 lines each?